src/v/transform/io.h

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "model/record.h"
+#include "model/record_batch_reader.h"
+#include "model/transform.h"
+
+namespace transform {
+namespace detail {
+
+/**
+ * A factory for sources and sinks.
+ *
+ * Use sink::factory or source::factory instead.
+ */
+template<typename T>
+class factory {
+public:
+    factory() = default;
+    factory(const factory&) = delete;
+    factory& operator=(const factory&) = delete;
+    factory(factory&&) = delete;
+    factory& operator=(factory&&) = delete;
+    virtual ~factory() = default;
+
+    virtual std::optional<std::unique_ptr<T>> create(model::ntp) = 0;
```

Comment:
does returning nullptr work as a stand-in for nullopt here?

Reply:
>Do you have thoughts on this pattern?

doesn't really matter much to me as long as we are consistent at least at the subsystem level.

> https://github.com/abseil/abseil-cpp/blob/b06ab1f3554eee58fa0eac046ea79f4b572fa502/absl/base/nullability.h

saw this yesterday when i was trying to upgrade our abseil dep.

Reply:
> saw this yesterday when i was trying to upgrade our abseil dep.

I'll probably switch over to this if we upgrade abseil.